File: sdk/kotlin/src/main/kotlin/com/privacyrpc/sdk/ZkCompression.kt
package com.privacyrpc.sdk

/**
 * ZK Compression Module for PrivacyRPC
 *
 * Tracks and manages ZK-compressed Solana RPC calls,
 * matching the ZK Compression panel in the Chrome extension.
 * Powered by Light Protocol.
 */
class ZkCompression {

    /** ZK Compression statistics */
    data class ZkStats(
        val compressedCalls: Long = 0,
        val regularCalls: Long = 0,
        val cacheHits: Long = 0,
        val estimatedSavings: Long = 0 // in lamports
    )

    private var stats = ZkStats()
    private val cache = mutableMapOf<String, CachedResult>()
    private val cacheTtlMs = 30_000L // 30 second cache

    private data class CachedResult(
        val result: String,
        val timestamp: Long
    )

    /** ZK Compression RPC methods (Light Protocol) */
    companion object {
        val ZK_METHODS = setOf(
            "getCompressedAccount",
            "getCompressedAccountsByOwner",
            "getCompressedBalance",
            "getCompressedBalanceByOwner",
            "getCompressedTokenAccountBalance",
            "getCompressedTokenAccountsByOwner",
            "getCompressedTokenAccountsByDelegate",
            "getCompressedTokenBalancesByOwner",
            "getCompressedMintTokenHolders",
            "getValidityProof",
            "getMultipleCompressedAccounts"
        )

        /** Estimated lamport savings per compressed call */
        private const val ESTIMATED_SAVINGS_PER_CALL = 1000L
    }

    /** Check if an RPC method is a ZK compression method */
    fun isZkMethod(method: String): Boolean {
        return method in ZK_METHODS
    }

    /** Record a ZK compressed call */
    fun recordCompressedCall(method: String, cacheKey: String? = null) {
        val now = System.currentTimeMillis()

        // Check cache
        if (cacheKey != null) {
            val cached = cache[cacheKey]
            if (cached != null && now - cached.timestamp < cacheTtlMs) {
                stats = stats.copy(
                    cacheHits = stats.cacheHits + 1,
                    estimatedSavings = stats.estimatedSavings + ESTIMATED_SAVINGS_PER_CALL
                )
                return
            }
        }

        stats = stats.copy(
            compressedCalls = stats.compressedCalls + 1,
            estimatedSavings = stats.estimatedSavings + ESTIMATED_SAVINGS_PER_CALL
        )
    }

    /** Record a regular (non-compressed) call */
    fun recordRegularCall() {
        stats = stats.copy(regularCalls = stats.regularCalls + 1)
    }

    /** Cache a ZK compression result */
    fun cacheResult(key: String, result: String) {
        cache[key] = CachedResult(result, System.currentTimeMillis())

        // Evict old entries
        val now = System.currentTimeMillis()
        cache.entries.removeIf { now - it.value.timestamp > cacheTtlMs }
    }

    /** Get current ZK stats */
    fun getStats(): ZkStats = stats

    /** Format savings for display */
    fun formatSavings(lamports: Long): String {
        return when {
            lamports < 1000 -> "$lamports lam"
            lamports < 1_000_000 -> "${lamports / 1000.0}K"
            else -> "${"%.4f".format(lamports / 1_000_000_000.0)} SOL"
        }
    }

    /** Reset stats */
    fun reset() {
        stats = ZkStats()
        cache.clear()
    }
}

File: sdk/swift/Sources/PrivacyRPCSDK/ZkCompression.swift
import Foundation

/// ZK Compression Module for PrivacyRPC
///
/// Tracks and manages ZK-compressed Solana RPC calls.
/// Powered by Light Protocol.
/// Matches the Chrome extension's ZK Compression panel.
public class ZkCompression {

    // MARK: - Models

    public struct ZkStats {
        public let compressedCalls: Int
        public let regularCalls: Int
        public let cacheHits: Int
        public let estimatedSavings: Int // in lamports

        public static let zero = ZkStats(compressedCalls: 0, regularCalls: 0, cacheHits: 0, estimatedSavings: 0)
    }

    // MARK: - Constants

    /// ZK Compression RPC methods (Light Protocol)
    public static let zkMethods: Set<String> = [
        "getCompressedAccount",
        "getCompressedAccountsByOwner",
        "getCompressedBalance",
        "getCompressedBalanceByOwner",
        "getCompressedTokenAccountBalance",
        "getCompressedTokenAccountsByOwner",
        "getCompressedTokenAccountsByDelegate",
        "getCompressedTokenBalancesByOwner",
        "getCompressedMintTokenHolders",
        "getValidityProof",
        "getMultipleCompressedAccounts"
    ]

    /// Estimated lamport savings per compressed call
    private static let estimatedSavingsPerCall = 1000

    // MARK: - State

    private var compressedCalls = 0
    private var regularCalls = 0
    private var cacheHits = 0
    private var estimatedSavings = 0

    private var cache: [String: CachedResult] = [:]
    private let cacheTtl: TimeInterval = 30 // 30 seconds

    private struct CachedResult {
        let result: String
        let timestamp: Date
    }

    public init() {}

    // MARK: - Public API

    /// Check if an RPC method is a ZK compression method
    public func isZkMethod(_ method: String) -> Bool {
        return Self.zkMethods.contains(method)
    }

    /// Record a ZK compressed call
    public func recordCompressedCall(method: String, cacheKey: String? = nil) {
        // Check cache
        if let key = cacheKey, let cached = cache[key] {
            if Date().timeIntervalSince(cached.timestamp) < cacheTtl {
                cacheHits += 1
                estimatedSavings += Self.estimatedSavingsPerCall
                return
            }
        }

        compressedCalls += 1
        estimatedSavings += Self.estimatedSavingsPerCall
    }

    /// Record a regular (non-compressed) call
    public func recordRegularCall() {
        regularCalls += 1
    }

    /// Cache a ZK compression result
    public func cacheResult(key: String, result: String) {
        cache[key] = CachedResult(result: result, timestamp: Date())

        // Evict old entries
        let now = Date()
        cache = cache.filter { now.timeIntervalSince($0.value.timestamp) < cacheTtl }
    }

    /// Get current ZK stats
    public func getStats() -> ZkStats {
        return ZkStats(
            compressedCalls: compressedCalls,
            regularCalls: regularCalls,
            cacheHits: cacheHits,
            estimatedSavings: estimatedSavings
        )
    }

    /// Format savings for display
    public func formatSavings(_ lamports: Int) -> String {
        if lamports < 1000 {
            return "\(lamports) lam"
        } else if lamports < 1_000_000 {
            return String(format: "%.1fK", Double(lamports) / 1000.0)
        } else {
            return String(format: "%.4f SOL", Double(lamports) / 1_000_000_000.0)
        }
    }

    /// Reset stats
    public func reset() {
        compressedCalls = 0
        regularCalls = 0
        cacheHits = 0
        estimatedSavings = 0
        cache.removeAll()
    }
}
